### PR TITLE
httpBackend updated to match Angular documentation

### PR DESCRIPTION
**Pull request by TonyCurtisLives**
Changes to httpBackend to conform to angular documentation here: https://code.angularjs.org/1.3.16/docs/api/ngMock/service/$httpBackend
Mainly add the option to pass a function in the url parameter for expect and when function flavors. Updated IRequestHandler respond function to return an IRequestHandler, updated overloads.
Used union types to reduce over 100 overloads to 19 signatures.
Added documentation and tests.

### CODE REVIEW

**Comment by joeskeen** [comment]
Looks good to me.

File: angularjs/angular-mocks-tests.ts [hunk 1]
```diff
@@ -291,15 +370,34 @@ requestHandler = httpBackendService.whenPUT(/test.local/, function (data: string
...
+requestHandler = httpBackendService.whenPUT((url: string) => { return true; }, { key: 'value' });
+requestHandler = httpBackendService.whenPUT((url: string) => { return true; }, { key: 'value' }, { header: 'value' });
 
 
 ///////////////////////////////////////
 // IRequestHandler
 ///////////////////////////////////////
+var expectedData = { key: 'value'};
 requestHandler.passThrough();
-requestHandler.respond(function () { });
```
Comment (daltin) [inline angularjs/angular-mocks-tests.ts]: According to current angular docs this empty function call should not be allowed, should pass 4 params and return boolean.

File: angularjs/angular-mocks-tests.ts [hunk 2]
```diff
@@ -291,15 +370,34 @@ requestHandler = httpBackendService.whenPUT(/test.local/, function (data: string
...
+requestHandler.passThrough().passThrough();
+requestHandler.respond((method, url, data, headers) => [404, 'data', { header: 'value' }, 'responseText']);
+requestHandler.respond((method, url, data, headers) => [404, 'data', { header: 'value' }, 'responseText']).respond({});
+requestHandler.respond((method, url, data, headers) => { return [404, { key: 'value' }, { header: 'value' }, 'responseText']; });
+requestHandler.respond('data');
+requestHandler.respond('data').respond({});
+requestHandler.respond(expectedData);
 requestHandler.respond({ key: 'value' });
 requestHandler.respond({ key: 'value' }, { header: 'value' });
-requestHandler.respond(404);
```
Comment (daltin) [inline angularjs/angular-mocks-tests.ts]: According to current angular docs the only non optional parameter is data (string).

File: angularjs/angular-mocks-tests.ts [hunk 3]
```diff
@@ -291,15 +370,34 @@ requestHandler = httpBackendService.whenPUT(/test.local/, function (data: string
...
+requestHandler = httpBackendService.whenPUT((url: string) => { return true; }, { key: 'value' }, { header: 'value' });
 
 
 ///////////////////////////////////////
 // IRequestHandler
 ///////////////////////////////////////
+var expectedData = { key: 'value'};
 requestHandler.passThrough();
-requestHandler.respond(function () { });
+requestHandler.passThrough().passThrough();
```
Comment (daltin) [inline angularjs/angular-mocks-tests.ts]: test fluent interface

File: angularjs/angular-mocks-tests.ts [hunk 4]
```diff
@@ -291,15 +370,34 @@ requestHandler = httpBackendService.whenPUT(/test.local/, function (data: string
...
+requestHandler.respond((method, url, data, headers) => [404, 'data', { header: 'value' }, 'responseText']);
+requestHandler.respond((method, url, data, headers) => [404, 'data', { header: 'value' }, 'responseText']).respond({});
```
Comment (daltin) [inline angularjs/angular-mocks-tests.ts]: test fluent interface

File: angularjs/angular-mocks-tests.ts [hunk 5]
```diff
@@ -291,15 +370,34 @@ requestHandler = httpBackendService.whenPUT(/test.local/, function (data: string
...
+requestHandler.respond((method, url, data, headers) => { return [404, { key: 'value' }, { header: 'value' }, 'responseText']; });
+requestHandler.respond('data');
+requestHandler.respond('data').respond({});
```
Comment (daltin) [inline angularjs/angular-mocks-tests.ts]: test fluent interface

File: angularjs/angular-mocks-tests.ts [hunk 6]
```diff
@@ -291,15 +370,34 @@ requestHandler = httpBackendService.whenPUT(/test.local/, function (data: string
...
+requestHandler.respond(expectedData);
 requestHandler.respond({ key: 'value' });
 requestHandler.respond({ key: 'value' }, { header: 'value' });
-requestHandler.respond(404);
+requestHandler.respond({ key: 'value' }, { header: 'value' }, 'responseText');
+requestHandler.respond(404, 'data');
+requestHandler.respond(404, 'data').respond({});
```
Comment (daltin) [inline angularjs/angular-mocks-tests.ts]: test fluent interface (3 overloads for respond so 3 fluent interface tests).

File: angularjs/angular-mocks.d.ts [hunk 7]
```diff
@@ -97,137 +97,208 @@ declare module angular {
...
+          * Verifies that all of the requests defined via the expect api were made. If any of the requests were not made, verifyNoOutstandingExpectation throws an exception.
+          */
         verifyNoOutstandingExpectation(): void;
+        
+        /**
+          * Verifies that there are no outstanding requests that need to be flushed.
+          */
         verifyNoOutstandingRequest(): void;
 
-        expect(method: string, url: string, data?: string, headers?: Object): mock.IRequestHandler;
```
Comment (daltin) [inline angularjs/angular-mocks.d.ts]: Had I used overloads for everything, adding the missing (url) => bool parameter option to all expect and when flavors we would have 140ish overloads, with documentation needed for each. I elected to use *unions*, which match what angular documented. This reduced the signature count to 16. I added documentation to all of those signatures. I don't know that I want to document 140 signatures.

File: angularjs/angular-mocks.d.ts [hunk 8]
```diff
@@ -97,137 +97,208 @@ declare module angular {
...
-
-        whenPUT(url: string, data?: string, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: string, data?: RegExp, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: string, data?: (data: string) => boolean, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: string, data?: Object, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: RegExp, data?: string, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: RegExp, data?: RegExp, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: RegExp, data?: (data: string) => boolean, headers?: Object): mock.IRequestHandler;
-        whenPUT(url: RegExp, data?: Object, headers?: Object): mock.IRequestHandler;
+        /**
```
Comment (daltin) [inline angularjs/angular-mocks.d.ts]: My 16 union based signatures with documentation which replace the roughly 140 signatures necessary via overloading.

File: angularjs/angular-mocks.d.ts [hunk 9]
```diff
@@ -97,137 +97,208 @@ declare module angular {
...
+            respond(status: number, data: string | Object, headers?: Object, responseText?: string): IRequestHandler;
+
+            /**
+              * Controls the response for a matched request using the HTTP status code 200 and supplied static data to construct the response. 
+              * Returns the RequestHandler object for possible overrides.
+              * @param data Data to add to the response.
+              * @param headers Headers object to add to the response.
+              * @param responseText Response text to add to the response.
+              */
+            respond(data: string | Object, headers?: Object, responseText?: string): IRequestHandler;
```
Comment (daltin) [inline angularjs/angular-mocks.d.ts]: unfortunately, as in the earlier version, with a signature that takes an object, and the rest optional, it gets confused with the single parameter function signature, so the IDE help is somewhat limited for the func signature

File: angularjs/angular-mocks.d.ts [hunk 10]
```diff
@@ -1,6 +1,6 @@
 // Type definitions for Angular JS 1.3 (ngMock, ngMockE2E module)
 // Project: http://angularjs.org
-// Definitions by: Diego Vilar <http://github.com/diegovilar>
+// Definitions by: Diego Vilar <http://github.com/diegovilar>, Tony Curtis <http://github.com/daltin>
```
Comment (daltin) [inline angularjs/angular-mocks.d.ts]: Not sure if I was supposed to add my name here for a major rewrite like this?

Comment (vvakame) [inline angularjs/angular-mocks.d.ts]: np

**Comment by vvakame** [comment]
@diegovilar could you review this PR?

**Comment by daltin** [comment]
@diegovilar could you please take a look at this pull request when you have time?

**Comment by vvakame** [comment]
if we can't get response in 3 days. please ping me again.

**Comment by daltin** [comment]
@vvakame Ok, looks like 3 days, could someone else possibly review this? I would greatly appreciate it!

**Comment by vvakame** [comment]
@daltin okay, It merged! thanks mate.

**Pull request merged**